File: butterflyui/src/lib/src/core/controls/layout/clip.dart
import 'package:flutter/material.dart';

import 'package:butterflyui_runtime/src/core/control_utils.dart';

Widget buildClipControl(
  Map<String, Object?> props,
  List<dynamic> rawChildren,
  Widget Function(Map<String, Object?> child) buildChild,
) {
  final child = _resolveChild(props, rawChildren, buildChild);
  final clipBehavior = _parseClip(props['clip_behavior']) ?? Clip.antiAlias;
  final shape = (props['shape'] ?? 'rrect').toString().toLowerCase();
  final radius = coerceDouble(props['radius'] ?? props['border_radius']) ?? 0.0;

  switch (shape) {
    case 'oval':
    case 'circle':
      return ClipOval(clipBehavior: clipBehavior, child: child);
    case 'rect':
    case 'rectangle':
      return ClipRect(clipBehavior: clipBehavior, child: child);
    default:
      return ClipRRect(
        clipBehavior: clipBehavior,
        borderRadius: BorderRadius.circular(radius),
        child: child,
      );
  }
}

Widget _resolveChild(
  Map<String, Object?> props,
  List<dynamic> rawChildren,
  Widget Function(Map<String, Object?> child) buildChild,
) {
  for (final raw in rawChildren) {
    if (raw is Map) {
      return buildChild(coerceObjectMap(raw));
    }
  }
  final propChild = props['child'];
  if (propChild is Map) {
    return buildChild(coerceObjectMap(propChild));
  }
  return const SizedBox.shrink();
}

Clip? _parseClip(Object? value) {
  switch (value?.toString().toLowerCase()) {
    case 'none':
      return Clip.none;
    case 'hardedge':
    case 'hard_edge':
      return Clip.hardEdge;
    case 'antialias':
    case 'anti_alias':
      return Clip.antiAlias;
    case 'antialiaswithsavelayer':
    case 'anti_alias_with_save_layer':
      return Clip.antiAliasWithSaveLayer;
  }
  return null;
}

File: butterflyui/sdk/python/packages/butterflyui/src/butterflyui/controls/layout/clip.py
from __future__ import annotations
from collections.abc import Mapping
from typing import Any
from .._shared import Component, merge_props

__all__ = ["Clip"]

class Clip(Component):
    """
    Clips its child to a shape such as a rounded rectangle or circle.

    The runtime selects a Flutter clip widget based on ``shape``: ``"rect"``
    maps to ``ClipRect``, ``"oval"`` maps to ``ClipOval``, and the default
    rounded-rect variant uses ``ClipRRect``. ``radius`` sets the corner radius.
    ``clip_behavior`` controls anti-aliasing: ``"hardEdge"``,
    ``"antiAlias"``, or ``"antiAliasWithSaveLayer"``.

    ```python
    import butterflyui as bui

    bui.Clip(
        bui.Image(src="avatar.png"),
        shape="oval",
    )
    ```

    Args:
        shape:
            Clip shape. Values: ``"rect"``, ``"oval"``, ``"rrect"`` (default).
        radius:
            Corner radius in logical pixels for rounded-rect clips.
        clip_behavior:
            Anti-aliasing mode. Values: ``"hardEdge"``, ``"antiAlias"``,
            ``"antiAliasWithSaveLayer"``.
    """

    control_type = "clip"

    def __init__(
        self,
        child: Any | None = None,
        *children: Any,
        shape: str | None = None,
        radius: float | None = None,
        clip_behavior: str | None = None,
        props: Mapping[str, Any] | None = None,
        style: Mapping[str, Any] | None = None,
        strict: bool = False,
        **kwargs: Any,
    ) -> None:
        merged = merge_props(
            props,
            shape=shape,
            radius=radius,
            clip_behavior=clip_behavior,
            **kwargs,
        )
        resolved_children = list(children)
        if child is not None:
            resolved_children.insert(0, child)
        super().__init__(*resolved_children, props=merged, style=style, strict=strict)
